Library/OcCpuLib/OcCpuLib.c

```diff
@@ -107,6 +88,161 @@ ScanThreadCount (
     }
   }
 
+  return Status;
+}
+
+STATIC
+EFI_STATUS
+ScanFrameworkMpServices (
+  IN  FRAMEWORK_EFI_MP_SERVICES_PROTOCOL  *FrameworkMpServices,
+  OUT OC_CPU_INFO                         *Cpu,
+  OUT UINTN                               *NumberOfProcessors,
+  OUT UINTN                               *NumberOfEnabledProcessors
+  )
+{
+  EFI_STATUS           Status;
+  UINTN                Index;
+  EFI_MP_PROC_CONTEXT  Context;
+  UINTN                ContextSize;
+
+  ASSERT (FrameworkMpServices != NULL);
+  ASSERT (Cpu != NULL);
+  ASSERT (NumberOfProcessors != NULL);
+  ASSERT (NumberOfEnabledProcessors != NULL);
+
+  ContextSize = sizeof (Context);
+
+  Status = FrameworkMpServices->GetGeneralMPInfo (
+    FrameworkMpServices,
+    NumberOfProcessors,
+    NULL,
+    NumberOfEnabledProcessors,
+    NULL,
+    NULL
+    );
+
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  if (*NumberOfProcessors == 0) {
+    return EFI_NOT_FOUND;
+  }
+
+  //
+  // This code assumes that all CPUs have same amount of cores and threads.
+  //
+  for (Index = 0; Index < *NumberOfProcessors; ++Index) {
+    Status = FrameworkMpServices->GetProcessorContext (
+      FrameworkMpServices,
+      Index,
+      &ContextSize,
```

Comment:
(1) must be performed within the loop every time before this function is called. Error conditions lead to nothing but a skip of the current protocol:
https://github.com/acidanthera/OcSupportPkg/blob/ae5a3c6d3c23cd98e2433731141b45d9ced7c821/Library/OcCpuLib/OcCpuLib.c#L143-L148
One of the error conditions is a too small buffer of the context struct:
https://github.com/acidanthera/EfiPkg/blob/af25c8b111eec06732f0a5a2a7c10f671cb23356/Include/Protocol/FrameworkMpService.h#L315-L317
This means if protocol _X_ requires a bigger context, it will update `ContextSize` to _Y_ and return `EFI_BUFFER_TOO_SMALL`. Then the loop will skip protocol _X_ and pretend to protocol _X_+1 that we have a context struct of size Y and not of size `sizeof (Context)`. We might want to actually use dynamic allocation to not needlessly fail, or at least print an error so we know when to update the logic.